src/04-modality-specific-files/03-electroencephalography.md

```diff
@@ -163,6 +163,12 @@ Example:
   "EEGPlacementScheme":"10 percent system",
   "EEGReference":"single electrode placed on FCz",
   "EEGGround":"placed on AFz",
+  "SoftwareFilters":{
```

Comment:
The spec says "List of temporal software filters applied.", and this is an object that does not have an inherent ordering. If ordering is important, than this should probably be lists of single-element objects. If the ordering is not important, then it can be kept as is, but I would suggest updating the language of the spec to change from "list", which has a meaning in JSON.

As an example of a list of single-element objects:

```JSON
  "SoftwareFilters": [
    {
      "Anti-aliasing filter": {
        "half-amplitude cutoff (Hz)": 500,
        "Roll-off": "6dB/Octave"
      }
    },
    {
      "Other filter": {
        "param1": "val1",
        "param2": "val2"
      }
    }
  ],
```

Reply:
So far, we have always dealt with objects, see also the [schema implemented in the validator](https://github.com/bids-standard/bids-validator/blob/4682c67f1ac55e87dd3205d7b81466fd376349a9/bids-validator/validators/json/schemas/eeg.json#L36-L41)

and the [examples in the starter-kit](https://github.com/bids-standard/bids-starter-kit/wiki/FAQ#how-to-format-hardware-and-software-filter-fields-in-a-json)

A change of language in the spec is thus warranted IMO.

Reply:
And order does not matter? Because software that parses JSON is not obligated to provide key-value pairs in any particular order, or preserve it when copying metadata forward to derivatives.

Reply:
This was never discussed in detail. I think in most use cases the order is irrelevant - but in cases where it IS relevant, perhaps a keyword `"order": "first" ...` could be used? --> This would however be more easily solved with a list that has a naturally defined order, as you mentioned above Chris.

cc @robertoostenveld for some third opinion

Reply:
tagging more EEG folks: Please read this thread and leave your opinion on how we want to format filters in BIDS. Estimated time: <5minutes @bids-standard/raw-electrophys

Reply:
1. I am not aware of "list" carrying a specific meaning in JSON. Could you point me to it?
2. Even if a list were recommended, I would argue that a singleton should be allowed. You could allow that in the validator using "anyOf" construct in the schema.

Reply:
1. As far as I understand it, [JavaScript Arrays](https://www.w3schools.com/js/js_arrays.asp) are a bit like Python lists.
1. yes, the idea is that:
    1. either we allow an object ("SoftwareFilters") which contains one object per filter used .... and these filter objected then have key value pairs for their parameters
    1. OR we allow a list ("SoftwareFilters"), which is empty if None were applied ... or it contains objects, each of which is corresponding to a filter. Then the order of objects in the list corresponds to the order in which the respective filters were applied


The second of the ideas above allows for an easy way to specify the order in which filters were used.

I think a practical solution would be to allow both variants. and IF the order of filters is relevant, then we RECOMMEND that users take the list of objects (solution 2) rather than the object of objects (solution 1) way.

Opinions?

Reply:
@jasmainak 

1. https://json.org/

> JSON is built on two structures:
> * A collection of name/value pairs. In various languages, this is realized as an object, record, struct, dictionary, hash table, keyed list, or associative array.
> * An ordered list of values. In most languages, this is realized as an array, vector, list, or sequence.

Their grammar does refer to the structure as an array, so I suppose there's some ambiguity as to whether "list" is a reserved word in that sense. But it does have a common computer-science meaning, which is referenced above.

2. Sure.

---

To the point, though: does ordering ever matter? If it does, then I would suggest that an actual list (or array) be used. If it does not, then we can refer to it as an object.

Reply:
> To the point, though: does ordering ever matter? If it does, then I would suggest that an actual list (or array) be used. If it does not, then we can refer to it as an object.

this is a signal processing question, as long as the filters are linear, then the order of application of the filters should not matter.

I am not sure about non-linear filters though.

Reply:
The example of an anti-aliasing filter without subsequent resampling does not make much sense for most acquisition systems. An anti aliasing filter would typically be implemented in hardware. I would prefer a better (i.e. a more common and more easily recognized) example.

Reply:
Example:

The Brain Products BrainAmp always samples the data at 5kHz, applying a 1kHz anti-aliasing filter. However, in the recording software settings, you can also select lower sampling rates (e.g., 1kHz).

--> If that is done, the software applies a "software filter" and downsamples the data.

This is reflected in the present example, which I find good :-)

If you prefer a better example, could you perhaps suggest one?